### ISSUE TITLE
freebsd 9.1 build failure

### ISSUE THREAD

**Issue by avsm** [janestreet/core_extended#2]
```
lib/extended_unix_stubs.c: In function 'quota_query':
lib/extended_unix_stubs.c:138: error: storage size of 'quota' isn't known
lib/extended_unix_stubs.c:144: error: 'Q_GETQUOTA' undeclared (first use in this function)
lib/extended_unix_stubs.c:147: warning: implicit declaration of function 'quotactl'
lib/extended_unix_stubs.c:138: warning: unused variable 'quota'
lib/extended_unix_stubs.c: In function 'quota_modify':
lib/extended_unix_stubs.c:177: error: storage size of 'quota' isn't known
lib/extended_unix_stubs.c:181: error: 'Q_SETQUOTA' undeclared (first use in this function)
lib/extended_unix_stubs.c:177: warning: unused variable 'quota'
Command exited with code 2.
### stderr ###
discover.sh:46:4: warning: #warning "cpp test --defined(LINUX_EXT)-- was false"
discover.sh:47:4: warning: #warning "Feature LINUX_EXT will not be availlable"
discover.sh:58:4: warning: #warning "cpp test --defined(RLIMIT_NICE)-- was false"
discover.sh:59:4: warning: #warning "Feature RLIMIT_NICE will not be availlable"
discover.sh:80:4: warning: #warning "cpp test --defined(_POSIX_SYNCHRONIZED_IO) && _POSIX_SYNCHRONIZED_IO > 0-- was false"
discover.sh:81:4: warning: #warning "Feature FDATASYNC will not be availlable"
getconf: no such configuration parameter `LFS64_CFLAGS'
E: Failure("Command ''/x/ocaml-4.00.1/bin/ocamlbuild' lib/libcore_extended_stubs.a lib/dllcore_extended_stubs.so lib/core_extended.cma lib/core_extended.cmxa lib/core_extended.a lib/core_extended.cmxs -tag debug' terminated with error code 10")
```

**Comment by dch** [comment]
Trying a minimal diff:
```
diff --git i/lib/extended_unix_stubs.c w/lib/extended_unix_stubs.c
index 40b4d10..20afbd0 100644
--- i/lib/extended_unix_stubs.c
+++ w/lib/extended_unix_stubs.c
@@ -11,8 +11,11 @@
 #include <errno.h>
 #include <string.h>
 #include <time.h>
+if defined(__FreeBSD__)
+# include <sys/types.h>
+# include <ufs/ufs/quota.h>
+#endif
 #include <sys/statvfs.h>
-#include <sys/quota.h>
 #include <sys/mount.h>

 #ifndef __USE_ISOC99
```

Isn't sufficient, still leaving us with:

```
grep -v '^#' setup.ml > setup_dev.ml
ocamlfind ocamlopt -o setup-dev.exe -linkpkg -package ocamlbuild,oasis.dynrun setup_dev.ml || 	  ocamlfind ocamlc -o setup-dev.exe -linkpkg -package ocamlbuild,oasis.dynrun setup_dev.ml || true
rm -f setup_dev.*
./setup-dev.exe -configure 
W: No replace section found in file Makefile
config/test.c:15:4: warning: "cpp test --defined(LINUX_EXT)-- was false" [-W#warnings]
#  warning "cpp test --defined(LINUX_EXT)-- was false"
   ^
config/test.c:16:4: warning: "Feature LINUX_EXT will not be availlable" [-W#warnings]
#  warning "Feature LINUX_EXT will not be availlable"
   ^
config/test.c:29:4: warning: "cpp test --defined(RLIMIT_NICE)-- was false" [-W#warnings]
#  warning "cpp test --defined(RLIMIT_NICE)-- was false"
   ^
config/test.c:30:4: warning: "Feature RLIMIT_NICE will not be availlable" [-W#warnings]
#  warning "Feature RLIMIT_NICE will not be availlable"
   ^
config/test.c:55:4: warning: "cpp test --defined(_POSIX_SYNCHRONIZED_IO) && _POSIX_SYNCHRONIZED_IO > 0-- was false" [-W#warnings]
#  warning "cpp test --defined(_POSIX_SYNCHRONIZED_IO) && _POSIX_SYNCHRONIZED_IO > 0-- was false"
   ^
config/test.c:56:4: warning: "Feature FDATASYNC will not be availlable" [-W#warnings]
#  warning "Feature FDATASYNC will not be availlable"
   ^
6 warnings generated.

Configuration: 
ocamlfind: ........................................... /home/dch/.opam/4.02.1/bin/ocamlfind
ocamlc: .............................................. /home/dch/.opam/4.02.1/bin/ocamlc.opt
ocamlopt: ............................................ /home/dch/.opam/4.02.1/bin/ocamlopt.opt
ocamlbuild: .......................................... /home/dch/.opam/4.02.1/bin/ocamlbuild
Package name: ........................................ core_extended
Package version: ..................................... 112.17.00
os_type: ............................................. Unix
system: .............................................. freebsd
architecture: ........................................ amd64
ccomp_type: .......................................... cc
ocaml_version: ....................................... 4.02.1
standard_library_default: ............................ /home/dch/.opam/4.02.1/lib/ocaml
standard_library: .................................... /home/dch/.opam/4.02.1/lib/ocaml
standard_runtime: .................................... /home/dch/.opam/4.02.1/bin/ocamlrun
bytecomp_c_compiler: ................................. cc -D_FILE_OFFSET_BITS=64 -D_REENTRANT -D_THREAD_SAFE -fPIC
native_c_compiler: ................................... cc -D_FILE_OFFSET_BITS=64 -D_REENTRANT -D_THREAD_SAFE
model: ............................................... default
ext_obj: ............................................. .o
ext_asm: ............................................. .s
ext_lib: ............................................. .a
ext_dll: ............................................. .so
default_executable_name: ............................. a.out
systhread_supported: ................................. true
Install architecture-independent files dir: .......... /usr/local
Install architecture-dependent files in dir: ......... $prefix
User executables: .................................... $exec_prefix/bin
System admin executables: ............................ $exec_prefix/sbin
Program executables: ................................. $exec_prefix/libexec
Read-only single-machine data: ....................... $prefix/etc
Modifiable architecture-independent data: ............ $prefix/com
Modifiable single-machine data: ...................... $prefix/var
Object code libraries: ............................... $exec_prefix/lib
Read-only arch-independent data root: ................ $prefix/share
Read-only architecture-independent data: ............. $datarootdir
Info documentation: .................................. $datarootdir/info
Locale-dependent data: ............................... $datarootdir/locale
Man documentation: ................................... $datarootdir/man
Documentation root: .................................. $datarootdir/doc/$pkg_name
HTML documentation: .................................. $docdir
DVI documentation: ................................... $docdir
PDF documentation: ................................... $docdir
PS documentation: .................................... $docdir
findlib_version: ..................................... 1.5.5
is_native: ........................................... true
suffix_program: ...................................... 
Remove a file.: ...................................... rm -f
Remove a directory.: ................................. rm -rf
Turn ocaml debug flag on: ............................ true
Turn ocaml profile flag on: .......................... false
Compiler support generation of .cmxs.: ............... true
OCamlbuild additional flags: ......................... 
Enable linux specific extensions: .................... false
Enable POSIX timers: ................................. true
Create documentations: ............................... true
Compile tests executable and library and run them: ... false
camlp4o: ............................................. /home/dch/.opam/4.02.1/bin/camlp4o
pkg_bigarray: ........................................ /home/dch/.opam/4.02.1/lib/ocaml
pkg_bin_prot: ........................................ /home/dch/.opam/4.02.1/lib/bin_prot
pkg_bin_prot_syntax: ................................. /home/dch/.opam/4.02.1/lib/bin_prot
pkg_comparelib_syntax: ............................... /home/dch/.opam/4.02.1/lib/comparelib
pkg_core: ............................................ /home/dch/.opam/4.02.1/lib/core
pkg_custom_printf: ................................... /home/dch/.opam/4.02.1/lib/custom_printf
pkg_custom_printf_syntax: ............................ /home/dch/.opam/4.02.1/lib/custom_printf
pkg_fieldslib: ....................................... /home/dch/.opam/4.02.1/lib/fieldslib
pkg_fieldslib_syntax: ................................ /home/dch/.opam/4.02.1/lib/fieldslib
pkg_pa_bench_syntax: ................................. /home/dch/.opam/4.02.1/lib/pa_bench
pkg_pa_ounit: ........................................ /home/dch/.opam/4.02.1/lib/pa_ounit
pkg_pa_ounit_syntax: ................................. /home/dch/.opam/4.02.1/lib/pa_ounit
pkg_pa_pipebang: ..................................... /home/dch/.opam/4.02.1/lib/pa_pipebang
pkg_pa_test: ......................................... /home/dch/.opam/4.02.1/lib/pa_test
pkg_pa_test_syntax: .................................. /home/dch/.opam/4.02.1/lib/pa_test
pkg_textutils: ....................................... /home/dch/.opam/4.02.1/lib/textutils
pkg_re2: ............................................. /home/dch/.opam/4.02.1/lib/re2
pkg_sexplib: ......................................... /home/dch/.opam/4.02.1/lib/sexplib
pkg_sexplib_syntax: .................................. /home/dch/.opam/4.02.1/lib/sexplib
pkg_sexplib_num: ..................................... /home/dch/.opam/4.02.1/lib/sexplib_num
pkg_unix: ............................................ /home/dch/.opam/4.02.1/lib/ocaml
pkg_threads: ......................................... /home/dch/.opam/4.02.1/lib/ocaml

./setup-dev.exe -build 
W: No replace section found in file Makefile
W: Cannot find source file matching module 'core_extended' in library core_extended
ocamlfind ocamlopt unix.cmxa -I /home/dch/.opam/4.02.1/lib/ocaml/ocamlbuild /home/dch/.opam/4.02.1/lib/ocaml/ocamlbuild/ocamlbuildlib.cmxa -linkpkg myocamlbuild.ml /home/dch/.opam/4.02.1/lib/ocaml/ocamlbuild/ocamlbuild.cmx -o myocamlbuild
/home/dch/.opam/4.02.1/bin/ocamlfind ocamlc -I lib -package core -thread -ccopt -Ilib -c lib/extended_int_stubs.c
mv extended_int_stubs.o lib/extended_int_stubs.o
/home/dch/.opam/4.02.1/bin/ocamlfind ocamlc -I lib -package core -thread -ccopt -Ilib -c lib/extended_linux_stubs.c
mv extended_linux_stubs.o lib/extended_linux_stubs.o
/home/dch/.opam/4.02.1/bin/ocamlfind ocamlc -I lib -package core -thread -ccopt -Ilib -c lib/extended_unix_stubs.c
+ /home/dch/.opam/4.02.1/bin/ocamlfind ocamlc -I lib -package core -thread -ccopt -Ilib -c lib/extended_unix_stubs.c
lib/extended_unix_stubs.c:14:1: error: expected identifier or '('
if defined(__FreeBSD__)
^
lib/extended_unix_stubs.c:17:2: error: #endif without #if
#endif
 ^
lib/extended_unix_stubs.c:147:55: error: no member named 'dqb_curbytes' in 'struct dqblk64'
  bytes_used = QUOTA_BYTES_PER_SPACE_UNIT * (int64_t) QUOTA_SPACE_USED(quota);
                                                      ^~~~~~~~~~~~~~~~~~~~~~~
lib/extended_unix_stubs.c:97:44: note: expanded from macro 'QUOTA_SPACE_USED'
#  define QUOTA_SPACE_USED(quota) ((quota).dqb_curbytes)
                                   ~~~~~~~ ^
3 errors generated.
Command exited with code 2.
E: Failure("Command ''/home/dch/.opam/4.02.1/bin/ocamlbuild' lib/libcore_extended_stubs.a lib/dllcore_extended_stubs.so lib/core_extended.cma lib/core_extended.cmxa lib/core_extended.a lib/core_extended.cmxs -use-ocamlfind -tag debug' terminated with error code 10")
Makefile:21: recipe for target 'build' failed
gmake: *** [build] Error 1
```

FWIW quota on FreeBSD is https://www.freebsd.org/cgi/man.cgi?query=quotactl&sektion=2 perhaps this is best handled by making it part of `lib/extended_linux.ml`?

**Comment by dch** [comment]
Just checking current release... 

```
#=== ERROR while installing core_extended.112.35.00 ===========================#
# opam-version         1.2.2
# os                   freebsd
# command              gmake
# path                 /home/dch/.opam/system/build/core_extended.112.35.00
# compiler             system (4.02.1)
# exit-code            2
# env-file             /home/dch/.opam/system/build/core_extended.112.35.00/core_extended-17805-e7966f.env
# stdout-file          /home/dch/.opam/system/build/core_extended.112.35.00/core_extended-17805-e7966f.out
# stderr-file          /home/dch/.opam/system/build/core_extended.112.35.00/core_extended-17805-e7966f.err
### stdout ###
# [...]
# /home/dch/.opam/system/bin/ocamlfind ocamlc -I src -package core -thread -ccopt -Isrc -c src/extended_linux_stubs.c
# mv extended_linux_stubs.o src/extended_linux_stubs.o
# /home/dch/.opam/system/bin/ocamlfind ocamlc -I src -package core -thread -ccopt -Isrc -c src/extended_unix_stubs.c
# + /home/dch/.opam/system/bin/ocamlfind ocamlc -I src -package core -thread -ccopt -Isrc -c src/extended_unix_stubs.c
# src/extended_unix_stubs.c:15:10: fatal error: 'sys/quota.h' file not found
# #include <sys/quota.h>
#          ^
# 1 error generated.
# Command exited with code 2.
# Makefile:21: recipe for target 'build' failed
### stderr ###
# [...]
# config/test.c:56:4: warning: "Feature FDATASYNC will not be availlable" [-W#warnings]
# #  warning "Feature FDATASYNC will not be availlable"
#    ^
# 6 warnings generated.
# W: Cannot find source file matching module 'core_extended' in library core_extended
# W: Cannot find source file matching module 'janecheck' in library janecheck
# W: Cannot find source file matching module 'janecheck_kernel' in library janecheck_kernel
# W: Cannot find source file matching module 'selector_lib' in library selector_lib
# E: Failure("Command ''/usr/local/bin/ocamlbuild' selector/src/selector_lib.cma selector/src/selector_lib.cmxa selector/src/selector_lib.a selector/src/selector_lib.cmxs janecheck_kernel/src/janecheck_kernel.cma janecheck_kernel/src/janecheck_kernel.cmxa janecheck_kernel/src/janecheck_kernel.a janecheck_kernel/src/janecheck_kernel.cmxs janecheck/src/janecheck.cma janecheck/src/janecheck.cmxa janecheck/src/janecheck.a janecheck/src/janecheck.cmxs src/libcore_extended_stubs.a src/dllcore_extended_stubs.so src/core_extended.cma src/core_extended.cmxa src/core_extended.a src/core_extended.cmxs -use-ocamlfind -tag debug' terminated with error code 10")
# gmake: *** [build] Error 1



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
The following actions were aborted
  ∗  install patdiff 112.24.00
The following actions failed
  ∗  install core_extended 112.35.00
No changes have been performed
```

#sadpanda

**Comment by dch** [comment]
The simple patch in ocaml/opam-repository#4870 does address this, would be great to include this in upstream instead of as an opam hack. Thanks!

**Comment by diml** [comment]
We internally included this patch, which will be part of the next release:

```
--- a/src/extended_unix_stubs.c
+++ b/src/extended_unix_stubs.c
@@ -12,7 +12,13 @@
 #include <string.h>
 #include <time.h>
 #include <sys/statvfs.h>
+
+#if defined (__FreeBSD__) || defined(__OpenBSD__)
+#include <ufs/ufs/quota.h>
+#else
 #include <sys/quota.h>
+#endif
+
 #include <sys/mount.h>

 #ifndef __USE_ISOC99
```

**Comment by seliopou** [comment]
I still get this error FreeBSD 10.0, core_extended.113.33.03:

```
#=== ERROR while installing core_extended.113.33.03 ===========================#
# opam-version 1.2.2 (391297300524717e1475deaea37bca287c7f5e36)
# os           freebsd
# command      gmake
# path         /home/vagrant/.opam/4.02.3/build/core_extended.113.33.03
# compiler     4.02.3
# exit-code    2
# env-file     /home/vagrant/.opam/4.02.3/build/core_extended.113.33.03/core_extended-88660-559acc.env
# stdout-file  /home/vagrant/.opam/4.02.3/build/core_extended.113.33.03/core_extended-88660-559acc.out
# stderr-file  /home/vagrant/.opam/4.02.3/build/core_extended.113.33.03/core_extended-88660-559acc.err
### stdout ###
# [...]
# + /home/vagrant/.opam/4.02.3/bin/ocamlfind ocamlc -g -I src -package core -thread -c src/extended_unix_stubs.c
# src/extended_unix_stubs.c: In function 'quota_query':
# src/extended_unix_stubs.c:100:43: error: 'struct dqblk64' has no member named 'dqb_curbytes'
#  #  define QUOTA_SPACE_USED(quota) ((quota).dqb_curbytes)
#                                            ^
# src/extended_unix_stubs.c:150:55: note: in expansion of macro 'QUOTA_SPACE_USED'
#    bytes_used = QUOTA_BYTES_PER_SPACE_UNIT * (int64_t) QUOTA_SPACE_USED(quota);
#                                                        ^
# Command exited with code 2.
# Makefile:15: recipe for target 'build' failed
### stderr ###
# W: Cannot find source file matching module 'core_extended' in library core_extended
# E: Failure("Command ''/home/vagrant/.opam/4.02.3/bin/ocamlbuild' src/libcore_extended_stubs.a src/dllcore_extended_stubs.so src/core_extended.cma src/core_extended.cmxa src/core_extended.a src/core_extended.cmxs selector/src/selector.cma selector/src/selector.cmxa selector/src/selector.a selector/src/selector.cmxs -use-ocamlfind -plugin-tags 'package(js-build-tools.ocamlbuild_goodies), package(ppx_driver.ocamlbuild)' -tag debug' terminated with error code 10")
# gmake: *** [build] Error 1
```

**Comment by seliopou** [comment]
Sorry, not the same error, but seems to be related.

**Comment by diml** [comment]
fixed by e3ba689e5365738786ab45f294c08c63cdcffe24
